From 56058a86e795905cd1f143edfa7fd5d030a4dde8 Mon Sep 17 00:00:00 2001
From: Guojing Yuan <guojing@google.com>
Date: Tue, 8 Jul 2025 18:56:34 +0000
Subject: [PATCH] [CDM] Limit NLS revoke to applicable roles.

Bug: 421919075
Flag: EXEMPT bugfix
Test: CTS
Change-Id: Icd3cdb3234a516ff7f752f39d7b1d8e6f5cde1ef
(cherry picked from commit 8ccb7103a2d58ec2cee0a1cf5f8c7fefa5e8e8ee)
---
 .../association/DisassociationProcessor.java  | 22 +++++++++++--------
 .../server/companion/utils/RolesUtils.java    | 11 ++++++++++
 2 files changed, 24 insertions(+), 9 deletions(-)

diff --git a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
index ac5917e5c0f1..1475117b8b28 100644
--- a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
+++ b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
@@ -20,6 +20,7 @@ import static android.app.ActivityManager.RunningAppProcessInfo.IMPORTANCE_VISIB
 import static android.companion.AssociationRequest.DEVICE_PROFILE_AUTOMOTIVE_PROJECTION;
 
 import static com.android.internal.util.CollectionUtils.any;
+import static com.android.server.companion.utils.RolesUtils.NLS_PROFILES;
 import static com.android.server.companion.utils.RolesUtils.removeRoleHolderForAssociation;
 
 import static java.util.concurrent.TimeUnit.DAYS;
@@ -139,15 +140,18 @@ public class DisassociationProcessor {
         // Revoke NLS if the last association has been removed for the package
         Binder.withCleanCallingIdentity(() -> {
             if (mAssociationStore.getAssociationsByPackage(userId, packageName).isEmpty()) {
-                Intent nlsIntent = new Intent(
-                        NotificationListenerService.SERVICE_INTERFACE);
-                List<ResolveInfo> matchedServiceList = mContext.getPackageManager()
-                        .queryIntentServicesAsUser(nlsIntent, /* flags */ 0, userId);
-                for (ResolveInfo service : matchedServiceList) {
-                    if (service.getComponentInfo().getComponentName().getPackageName()
-                            .equals(packageName)) {
-                        mNotificationManager.setNotificationListenerAccessGranted(
-                                service.getComponentInfo().getComponentName(), false);
+                if (association.getDeviceProfile() != null
+                        && NLS_PROFILES.contains(association.getDeviceProfile())) {
+                    Intent nlsIntent = new Intent(
+                            NotificationListenerService.SERVICE_INTERFACE);
+                    List<ResolveInfo> matchedServiceList = mContext.getPackageManager()
+                            .queryIntentServicesAsUser(nlsIntent, /* flags */ 0, userId);
+                    for (ResolveInfo service : matchedServiceList) {
+                        if (service.getComponentInfo().getComponentName().getPackageName()
+                                .equals(packageName)) {
+                            mNotificationManager.setNotificationListenerAccessGranted(
+                                    service.getComponentInfo().getComponentName(), false, false);
+                        }
                     }
                 }
             }
diff --git a/services/companion/java/com/android/server/companion/utils/RolesUtils.java b/services/companion/java/com/android/server/companion/utils/RolesUtils.java
index dd12e0406089..5d2c05549616 100644
--- a/services/companion/java/com/android/server/companion/utils/RolesUtils.java
+++ b/services/companion/java/com/android/server/companion/utils/RolesUtils.java
@@ -17,6 +17,10 @@
 package com.android.server.companion.utils;
 
 import static android.app.role.RoleManager.MANAGE_HOLDERS_FLAG_DONT_KILL_APP;
+import static android.companion.AssociationRequest.DEVICE_PROFILE_APP_STREAMING;
+import static android.companion.AssociationRequest.DEVICE_PROFILE_COMPUTER;
+import static android.companion.AssociationRequest.DEVICE_PROFILE_GLASSES;
+import static android.companion.AssociationRequest.DEVICE_PROFILE_WATCH;
 
 import android.annotation.NonNull;
 import android.annotation.SuppressLint;
@@ -29,6 +33,7 @@ import android.os.UserHandle;
 import android.util.Slog;
 
 import java.util.List;
+import java.util.Set;
 import java.util.function.Consumer;
 
 /** Utility methods for accessing {@link RoleManager} APIs. */
@@ -37,6 +42,12 @@ public final class RolesUtils {
 
     private static final String TAG = "CDM_RolesUtils";
 
+    public static final Set<String> NLS_PROFILES = Set.of(
+            DEVICE_PROFILE_WATCH,
+            DEVICE_PROFILE_GLASSES,
+            DEVICE_PROFILE_APP_STREAMING,
+            DEVICE_PROFILE_COMPUTER);
+
     /**
      * Check if the package holds the role.
      */
-- 
2.52.0.rc1.455.g30608eb744-goog

